File: 2005/Windows SDK NetFx 3.0/Technologies/NETCF/ManagedDirect3DMobileBillboardSample/VB/Billboard/GraphicsMesh.vb
'---------------------------------------------------------------------
'  This file is part of the Microsoft .NET Framework SDK Code Samples.
' 
'  Copyright (C) Microsoft Corporation.  All rights reserved.
' 
'This source code is intended only as a supplement to Microsoft
'Development Tools and/or on-line documentation.  See these other
'materials for detailed information regarding Microsoft code samples.
' 
'THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
'KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'PARTICULAR PURPOSE.
'---------------------------------------------------------------------


Imports System
Imports System.IO
Imports Microsoft.WindowsMobile.DirectX
Imports Microsoft.WindowsMobile.DirectX.Direct3D
Imports System.Text
Imports System.Reflection

Namespace Microsoft.Samples.MD3DM

    'Encapsulates some basic functionality to load and render meshes
    Class GraphicsMesh

        ' The rendering device
        Private device As Device

        ' The mesh this class loads and renders
        Private meshValue As Mesh = Nothing

        ' Materials for our mesh
        Private meshMaterials() As Material

        ' Textures for our mesh
        Private meshTextures() As Texture

        ' The assembly which has the resources embedded
        Private embeddedResourceAssembly As [Assembly]

        ' The string which will be prepended to filenames
        ' to produce valid resource names
        Private resourcePrefix As String

        ' Name of the mesh
        Private meshName As String

        ' The vertex format to load the mesh with
        Private vf As VertexFormats

        ' True if a vertex format has been set
        Private vfSet As Boolean = False
        
        Public ReadOnly Property Mesh() As Mesh
            Get
                Return meshValue
            End Get
        End Property        

        'Creates a new mesh
        Public Sub Create(ByVal device As Device, _
            ByVal resourcePrefix As String, _
            ByVal filename As String, ByVal [assembly] As [Assembly])
            Me.device = device
            embeddedResourceAssembly = [assembly]
            Me.meshName = filename
            Me.resourcePrefix = resourcePrefix

        End Sub 'Create
        
        'Sets the mesh vertex format
        Public Sub SetVertexFormat(ByVal format As VertexFormats) 
            vf = format
            vfSet = True
        End Sub 'SetVertexFormat
        
        ' Restores any needed data after a device reset
        Public Sub RestoreDeviceObjects(ByVal obj As Object, _
            ByVal eventg As EventArgs) 
            Dim textureFilenames As String() = Nothing
            
            meshValue = MeshLoader.LoadMesh(device, _
                embeddedResourceAssembly.GetManifestResourceStream( _
                resourcePrefix + meshName), _
                MeshFlags.SystemMemory, meshMaterials, textureFilenames)
            
            meshTextures = New Texture(meshMaterials.Length-1) {}
            Dim i As Integer
            For i = 0 To meshMaterials.Length-1
                ' Set the ambient color for the material 
                ' (D3DX does not do this)
                meshMaterials(i).Ambient = meshMaterials(i).Diffuse
                
                ' Create the texture
                If Not (textureFilenames(i) Is Nothing) Then
                    meshTextures(i) = TextureLoader.FromStream(device, _
                        embeddedResourceAssembly.GetManifestResourceStream( _
                        resourcePrefix + textureFilenames(i)))
                End If
            Next i
            
            If vf <> meshValue.VertexFormat AndAlso vfSet Then
                Dim temp As Mesh
                temp = meshValue.Clone(0, vf, meshValue.Device)
                meshValue.Dispose()
                meshValue = temp
            End If
        End Sub 'RestoreDeviceObjects
         
        'Renders the mesh
        Overloads Public Sub Render(ByVal canDrawOpaque As Boolean, _
            ByVal canDrawAlpha As Boolean) 
            Dim rs As RenderStateManager = device.RenderState
            ' Frist, draw the subsets without alpha
            If canDrawOpaque Then
                Dim i As Integer
                For i = 0 To meshMaterials.Length-1
                    If Not canDrawAlpha or meshMaterials(i).Diffuse.A >= _
                        &HFF Then
                        ' Set the material and texture for this subset
                        device.Material = meshMaterials(i)
                        If Not (meshTextures(i) Is Nothing) Then
                            device.SetTexture(0, meshTextures(i))
                        End If
                    
                        ' Draw the mesh subset
                        meshValue.DrawSubset(i)
                    End If 
                Next i
            End If
            
            ' Then, draw the subsets with alpha
            If canDrawAlpha Then
                ' Enable alpha blending
                rs.AlphaBlendEnable = True
                rs.SourceBlend = Blend.SourceAlpha
                rs.DestinationBlend = Blend.InvSourceAlpha
                Dim i As Integer
                For i = 0 To meshMaterials.Length-1
                    If meshMaterials(i).Diffuse.A <> &HFF Then
                        ' Set the material and texture for this subset
                        device.Material = meshMaterials(i)
                        If Not (meshTextures(i) Is Nothing) Then
                            device.SetTexture(0, meshTextures(i))
                        End If
                    
                        ' Draw the mesh subset
                        meshValue.DrawSubset(i)
                    End If 
                Next i
                ' Restore state
                rs.AlphaBlendEnable = False
            End If
        End Sub 'Render
        
        ' Renders the mesh
        Overloads Public Sub Render() 
            Me.Render(True, True)
        
        End Sub 'Render
    End Class 'GraphicsMesh 
End Namespace

File: 2005/Windows SDK NetFx 3.0/Technologies/NETCF/ManagedDirect3DMobileStencilBufferStencilDepthSample/CS/StencilBufferStencilDepth/GraphicsMesh.cs
//---------------------------------------------------------------------
//  This file is part of the Microsoft .NET Framework SDK Code Samples.
// 
//  Copyright (C) Microsoft Corporation.  All rights reserved.
// 
//This source code is intended only as a supplement to Microsoft
//Development Tools and/or on-line documentation.  See these other
//materials for detailed information regarding Microsoft code samples.
// 
//THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
//KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
//IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//PARTICULAR PURPOSE.
//---------------------------------------------------------------------

using System;
using System.IO;
using Microsoft.WindowsMobile.DirectX;
using Microsoft.WindowsMobile.DirectX.Direct3D;
using System.Text;
using System.Reflection;


namespace Microsoft.Samples.MD3DM
{

    /// <summary>
    /// A class for loading and displaying meshes
    /// </summary>
    class GraphicsMesh
    {
        // The rendering device
        Device device;

        // The mesh this class loads and renders
        Mesh meshValue = null;

        // Materials for our mesh
        Material[] meshMaterials; 
        
        // Textures for our mesh
        Texture[] meshTextures; 
        
        // The assembly which has the resources embedded
        Assembly embeddedResourceAssembly;
        
        // name of the mesh
        string meshName;

		// the name that must be prepended to filenames
		// in order to reference the resources from the
		// assembly
		string resourcePrefix;
        
        // the vertex format to load the mesh with
        VertexFormats vf;

        // true if a vertex format has been set
        bool vfSet = false;

        public Mesh Mesh
        {
            get { return meshValue; }
        }

        /// <summary>
        /// Creates a new mesh
        /// </summary>
        public void Create(Device device, string resourcePrefix,
			string filename, Assembly assembly)
        {
            this.device = device;
            embeddedResourceAssembly = assembly;
            this.meshName = filename;
			this.resourcePrefix = resourcePrefix;
        }

        /// <summary>
        /// Sets the desired vertex format for the mesh
        /// </summary>
        public void SetVertexFormat(VertexFormats format)
        {
            vf = format;
            vfSet = true;
        }

        /// <summary>
        /// The device exists, but may have just been Reset().  Resources
        /// and any other device state that persists during
        /// rendering should be set here.  Render states, matrices, textures,
        /// etc., that don't change during rendering can be set once here to
        /// avoid redundant state setting during Render() or FrameMove().
        /// </summary>
        public void RestoreDeviceObjects(object obj, EventArgs eventg)
        {
            string[] textureFilenames = null;

            meshValue = MeshLoader.LoadMesh(device, 
                embeddedResourceAssembly.GetManifestResourceStream(
				resourcePrefix + meshName), 
                MeshFlags.SystemMemory, out meshMaterials, 
                out textureFilenames);

            meshTextures  = new Texture[meshMaterials.Length];
            for( int i=0; i<meshMaterials.Length; i++ )
            {
                // Set the ambient color for the material 
                // (D3DX does not do this)
                meshMaterials[i].Ambient = meshMaterials[i].Diffuse;

                // Create the texture
                if (textureFilenames[i] != null)
                {
                    meshTextures[i] = TextureLoader.FromStream(device, 
                        embeddedResourceAssembly.GetManifestResourceStream(
                        resourcePrefix + textureFilenames[i]));
                }
            }
            
            // if the mesh was not loaded in the desired vertex format
            // it can be changed here
            if ((vf != meshValue.VertexFormat) && vfSet)
            {
                Mesh temp = meshValue.Clone(0, vf, meshValue.Device);
                meshValue.Dispose();
                meshValue = temp;
            }
        }
    
        /// <summary>
        /// Renders the mesh
        /// </summary>
        /// <param name="canDrawOpaque">Enables drawing opaque portions
        /// of the mesh</param>
        /// <param name="canDrawAlpha">Enables drawing alpha blended portions
        /// of the mesh</param>
        public void Render(bool canDrawOpaque, bool canDrawAlpha)
        {
            RenderStateManager rs = device.RenderState;
            // Frist, draw the subsets without alpha
            if (canDrawOpaque)
            {
                for( int i=0; i<meshMaterials.Length; i++ )
                {
                    if (canDrawAlpha)
                    {
                        if (meshMaterials[i].Diffuse.A < 0xff)
                            continue;
                    }
                    // Set the material and texture for this subset
                    device.Material = meshMaterials[i];

                    if( meshTextures[i] != null)
                        device.SetTexture(0, meshTextures[i]);

                    // Draw the mesh subset
                    meshValue.DrawSubset(i);
                }
            }

            // Then, draw the subsets with alpha
            if (canDrawAlpha)
            {
                // Enable alpha blending
                rs.AlphaBlendEnable = true;
                rs.SourceBlend = Blend.SourceAlpha;
                rs.DestinationBlend = Blend.InvSourceAlpha;
                for( int i=0; i<meshMaterials.Length; i++ )
                {
                    if (meshMaterials[i].Diffuse.A == 0xff)
                        continue;

                    // Set the material and texture for this subset
                    device.Material = meshMaterials[i];

                    if(meshTextures[i] != null)
                        device.SetTexture(0, meshTextures[i]);

                    // Draw the mesh subset
                    meshValue.DrawSubset(i);
                }
                // Restore state
                rs.AlphaBlendEnable = false;
            }
        }

        /// <summary>
        /// Renders the mesh
        /// </summary>
        public void Render()
        {
            this.Render(true, true);
        }
    }
}
